### ISSUE TITLE
Unable to set node affinity using "node.unique.name" attribute

### ISSUE THREAD

**Issue by radcool** [hashicorp/nomad#6926]
Nomad v0.10.1 (0d4e5d949fe073c47a947ea36bfef31a3c49224f)
CentOS 7 (3.10.0-957.21.3.el7.x86_64)

Hi,

I'm trying to set node affinity for different unrelated jobs so that I can manually spread them across three different Nomad clients (basically I'm trying to counter bin-packing).

Here is the client list for my Nomad cluster:

```
[root@arch ~]# nomad node status
ID        DC   Name     Class   Drain  Eligibility  Status
136bbdea  dc1  nomad09  <none>  false  eligible     ready
4e60814e  dc1  nomad08  <none>  false  eligible     ready
101028da  dc1  nomad07  <none>  false  eligible     ready
a21d9923  dc1  nomad06  <none>  false  eligible     ready
e74904d9  dc1  nomad05  <none>  false  eligible     ready
5c1e3b13  dc1  nomad04  <none>  false  eligible     ready
```

Note that only **nomad07**, **nomad08** and **nomad09** will be possible candidates for placement as they are the only nodes offering the 'raw_exec' driver, which my jobs are configured to use.

I've tried setting the affinity by using the `node.unique.name` attribute at the task level:

```
affinity {
  attribute = "${node.unique.name}"
  value     = "nomad09"
  weight    = 100
}
```

Unfortunately the task does not get scheduled on the desired node and there is no indication that the attribute is taken into consideration in the placement metric calculation:

```
[root@arch ~]# nomad status -verbose e01cc4b5

[...]

Placement Metrics
  * Constraint "missing drivers" filtered 3 nodes
Node                                  node-reschedule-penalty  binpack  job-anti-affinity  final score
101028da-bbf0-2873-ccbf-edc50630a0d9  0                        0.631    0                  0.631
136bbdea-be66-86a4-e4fd-b1fd87eede24  0                        0.205    0                  0.603
4e60814e-9caf-88a3-4b52-71bfe6975784  0                        0.402    0                  0.402
```

However, if I change my affinity stanza to use the `node.unique.id` attribute:

```
affinity {
  attribute = "${node.unique.id}"
  value     = "101028da-bbf0-2873-ccbf-edc50630a0d9"
  weight    = 100
}
```

then the scheduler _does_ take the node affinity into account:

```
[root@arch ~]# nomad status -verbose 9f5c205b

Placement Metrics
  * Constraint "missing drivers" filtered 3 nodes
Node                                  job-anti-affinity  node-reschedule-penalty  node-affinity  binpack  final score
101028da-bbf0-2873-ccbf-edc50630a0d9  0                  0                        1              0.532    0.766
4e60814e-9caf-88a3-4b52-71bfe6975784  0                  0                        0              0.471    0.471
136bbdea-be66-86a4-e4fd-b1fd87eede24  0                  0                        0              0.205    0.205
```

Why does affinity not seem to work when using the `node.unique.name` attribute but it does when using the `node.unique.id attribute`? Am I doing something wrong? Or is this a bug?

Thanks,
-Martin

**Comment by tgross** [comment]
Hi @radcool! Thanks for reporting this!

Interestingly, I wasn't able to reproduce this behavior on either 0.10.1 or the current master with the following job:

```hcl
job "example" {
  datacenters = ["dc1"]

  group "cache" {
    task "redis" {
      driver = "docker"

      affinity {
        attribute = "${node.unique.name}"
        value     = "client2"
        weight    = 100
      }

      config {
        image = "redis:3.2"

        port_map {
          db = 6379
        }
      }

      resources {
        cpu    = 500
        memory = 256

        network {
          mbits = 10
          port  "db"  {}
        }
      }
    }
  }
}
```


```
$ nomad node status
ID        DC   Name     Class   Drain  Eligibility  Status
8a4c0f3d  dc1  client2  <none>  false  eligible     ready
556d2e55  dc1  client1  <none>  false  eligible     ready

$ nomad alloc status d3e5d5eb
...
Placement Metrics
Node                                  binpack  job-anti-affinity  node-reschedule-penalty  node-affinity  final score
8a4c0f3d-b483-cb24-a017-579af0135158  0.176    0                  0                        1              0.588
556d2e55-622f-6e98-acfa-f25c12b9cd2a  0.176    0                  0                        0              0.176
```

Given the error message `* Constraint "missing drivers" filtered 3 nodes`, that suggests that the scheduler looked at the 3 nodes (which are exactly the 3 nodes you're saying have the `raw_exec` driver enabled), but don't. Can you check the detailed node status for one of those nodes? The drivers section should look something like (the "driver must run as root" won't be there if you're running Nomad as root):

```
$ nomad node status -verbose 101028da
ID          = 101028da-bbf0-2873-ccbf-edc50630a0d9
...
Drivers
Driver       Detected  Healthy  Message                  Time
docker       true      true     Healthy                  2020-01-10T13:32:14Z
exec         false     false    Driver must run as root  2020-01-10T13:32:14Z
java         false     false    Driver must run as root  2020-01-10T13:32:14Z
mock_driver  true      true     Healthy                  2020-01-10T13:32:14Z
qemu         true      true     Healthy                  2020-01-10T13:32:14Z
raw_exec     true      true     Healthy                  2020-01-10T13:32:14Z
rkt          false     false    Driver must run as root  2020-01-10T13:32:14Z
```

**Comment by radcool** [comment]
@tgross I ran `nomad node status -verbose XXXXXX` on all three `raw_exec` nodes and they are all `Healthy`.

That being said, to continue testingI tried creating a simple job spec with one group and a single task whose command is simply `sleep 60` and with a node affinity using the `node.unique.name` attribute. Lo and behold, that one actually got placed according to affinity and I confirmed that the `node-affinity` was indeed `1` for that job in the ouptut of `nomad alloc status`!

Out of curiosity, I then tried modifying one of my spec files from ealier, one of the ones where `node.unique.name` didn't seem to take effect, and this time, it actually worked!

So I don't really know what to make of this and I don't know what to say. I'm pretty sure I was testing correctly previously, but I'm now unable to reproduce that earlier, buggy behavior...
